src/TypeInference/DataType.mli

```diff
@@ -9,10 +9,11 @@
 open Common
 
 (** Open ADT making its constructors immediately available. *)
-val open_data : Env.t -> Env.adt_info -> Env.t
+val open_data : Env.t -> public:bool -> Module.adt_info -> Env.t
 
 (** Check non-recursive definition of ADT *)
-val check_data_def : Env.t -> S.data_def -> Env.t * T.data_def
+val check_data_def : Env.t -> public:bool -> S.data_def -> Env.t * T.data_def
 
 (** Check mutually recursive definitions of ADTs *)
-val check_rec_data_defs : Env.t -> S.data_def list -> Env.t * T.data_def list
+val check_rec_data_defs :
+  Env.t -> public:bool -> S.data_def list -> Env.t * T.data_def list
```

Comment:
The `public` flag is shared between all mutually recursive types. Should be possible to define public and private types that are mutually recursive? For regular value, such a restriction might be annoying, but not crucial. Mutually recursive values might be private, but some of them might have public aliases. But we do not have type aliases.

BTW, should we have `abstract` visibility flag for datatypes? And should this potential `abstract` flag have something in common with existential types?